libraries/egenesis/embed_genesis.cpp

```diff
@@ -0,0 +1,288 @@
+                  char dg[3];
+                  dg[0] = '0' + ((c >> 6) & 3);
+                  dg[1] = '0' + ((c >> 3) & 7);
+                  dg[2] = '0' + ((c     ) & 7);
+                  int start = (dg[0] == '0' ? (dg[1] == '0' ? 2 : 1) : 0);
+                  dest.append( dg+start, 3-start );
+            }
+            // clang-format on
+        }
+        dest.append("\"");
+        needs_comma = true;
+        row++;
+    }
+}
+
+struct egenesis_info
+{
+    fc::optional<sc::genesis_state_type> genesis;
+    fc::optional<sp::chain_id_type> chain_id;
+    fc::optional<std::string> genesis_json;
+    fc::optional<fc::sha256> genesis_json_hash;
+    fc::optional<std::string> genesis_json_array;
+    int genesis_json_array_width;
+    int genesis_json_array_height;
+
+    void fillin()
+    {
+        // must specify either genesis_json or genesis
+        if (genesis.valid())
+        {
+            if (!genesis_json.valid())
+                // If genesis_json not exist, generate from genesis
+                genesis_json = fc::json::to_string(*genesis);
+        }
+        else if (genesis_json.valid())
+        {
+            // If genesis not exist, generate from genesis_json
+            genesis = fc::json::from_string(*genesis_json).as<sc::genesis_state_type>();
+        }
+        else
+        {
+            // Neither genesis nor genesis_json exists, crippled
+            std::cerr << "embed_genesis:  Need genesis or genesis_json\n";
+            exit(1);
+        }
+
+        if (!genesis_json_hash.valid())
+            genesis_json_hash = fc::sha256::hash(*genesis_json);
+
+        if (!chain_id.valid())
+            chain_id = genesis_json_hash;
+
+        // init genesis_json_array from genesis_json
+        if (!genesis_json_array.valid())
+        {
+            genesis_json_array = std::string();
+            // TODO: gzip
+            int width = 40;
+            convert_to_c_array(*genesis_json, *genesis_json_array, width);
+            int height = (genesis_json->length() + width - 1) / width;
+            genesis_json_array_width = width;
+            genesis_json_array_height = height;
+        }
+    }
+};
+
+void load_genesis(const boost::program_options::variables_map& options, egenesis_info& info)
+{
+    if (options.count("genesis-json"))
+    {
+        fc::path genesis_json_filename = get_path(options, "genesis-json");
+        std::cout << "embed_genesis:  Reading genesis from file " << genesis_json_filename.preferred_string() << "\n";
+        info.genesis_json = std::string();
+        fc::read_file_contents(genesis_json_filename, *info.genesis_json);
+    }
+    else
+    {
+        sc::genesis_state_type genesis;
+        scorum::chain::utils::generate_default_genesis_state(genesis);
+        info.genesis = genesis;
+    }
+}
+
+void write_to_file(const fc::path& path, const std::string& content)
+{
+    fc::ofstream outfile(path);
+    outfile.write(content.c_str(), content.size());
```

Comment:
What will be if path is not valid or error will occure during writing to file?

Reply:
exception

Reply:
I will add checks